File: 24-adv-ds1/4-fenwick-trees/main.py
# Fenwick Trees - Python example

class FenwickTree:
    def __init__(self, size):
        self.size = size
        self.tree = [0] * (size + 1)

    def update(self, idx, delta):
        while idx <= self.size:
            self.tree[idx] += delta
            idx += idx & (-idx)

    def query(self, idx):
        result = 0
        while idx > 0:
            result += self.tree[idx]
            idx -= idx & (-idx)
        return result

if __name__ == "__main__":
    ft = FenwickTree(6)
    data = [1, 3, 5, 7, 9, 11]
    for i, val in enumerate(data, 1):
        ft.update(i, val)

    print(ft.query(3))  # Output: 9 (sum of first 3 elements)
    ft.update(3, 2)    # Increment 3rd element by 2
    print(ft.query(3))  # Output: 11

File: 24-adv-ds1/4-fenwick-trees/main.java
// Fenwick Trees - Java example

class Main {
    static class FenwickTree {
        int[] tree;
        int size;

        FenwickTree(int n) {
            size = n;
            tree = new int[n + 1];
        }

        void update(int idx, int delta) {
            while (idx <= size) {
                tree[idx] += delta;
                idx += idx & (-idx);
            }
        }

        int query(int idx) {
            int result = 0;
            while (idx > 0) {
                result += tree[idx];
                idx -= idx & (-idx);
            }
            return result;
        }
    }

    public static void main(String[] args) {
        int n = 6;
        FenwickTree ft = new FenwickTree(n);
        int[] data = {1, 3, 5, 7, 9, 11};
        for (int i = 1; i <= n; i++) {
            ft.update(i, data[i - 1]);
        }

        System.out.println(ft.query(3));  // Output: 9
        ft.update(3, 2);
        System.out.println(ft.query(3));  // Output: 11
    }
}
